ADBDEV-4207 Fix partition selection during DML execution when dropped columns are present

bimboterminator1 opened the pull request:
Fix partition selection during DML execution when dropped columns are present

Current partitioning design faced a problem, which is connected with the
differences in physical attribute layouts of the child partition and the parent
relation in case when the parent relation has dropped columns and the newly
added child partition has not, and vice versa.

When executing modifying DML operations on a leaf partition of a partitioned
table, the executor anyway performs the partition selection in case of INSERT,
UPDATE for legacy planner and in case of INSERT, UPDATE, DELETE for ORCA
optimizer. This is done in `selectPartition` function, which is called from
`slot_get_partition` function (in case of modifying DML commands). This procedure 
is done based on 1) attribute values of the tuple, which have been retrieved
from the respective subplan of ModifyTable (DML) node and whose attribute
structure corresponds the leaf partition's. 2) the root partition's attribute
numbers and partition rules, which are based on the parent (root) relation's
tuple descriptor. Thus, **if the parent relation's tuple descriptor contained
dropped columns and leaf partition's did not** (or visa versa, the leaf partition
had dropped columns and parent had not), the partition selection could go wrong
and lead to various execution errors.

Let's consider the following case. The partitioned table was created, and at
one moment it was decided to drop several irrelevant columns from it. After that
some new partitions were added or some were exchanged with new relations. In
this situation these newly added (exchanged) partitions don't have dropped
columns in it's structure, but the original parent relation has such dropped
attributes. Then the INSERT command into the new leaf partition is executed. In
this case for both planners the `ExecInsert` function is called, which then calls
`slot_get_partition` function in order to validate that the tuple values
correspond to specified leaf partition. And here the tuple descriptor does not
have dropped attributes, but the **partition selection is performed from parent's
view**, which have those attributes, and the parent's partitioning attribute
numbers are used to validate the tuple. Because of that the partition selection
procedure could either select wrong target partition (the wrong attribute was
taken from the slot values to check the partition rule) or fail to find the
partition at all, becasue none of the partition rules were satisfied. The same
issue occured when we didn't drop anything from parent relation, however we
exhanged a partition to the one with dropped columns. When inserting into
exchanged leaf partition the partition selection could also go wrong by the same
reasons (selection is performed from parent's view).

Similar issue occured for both planners in other commands, when partition
selection was performed. And in order to prevent the potentially wrong partition
selection, when modifying the leaf partition, this patch proposes the
following solution.

The legacy planner have `checkPartitionUpdate` function, which is called in UPDATE
cases and which checks that the tuple does not switch its partition at
update. And to our astonishment this function is able to handle the case when
the partition being modified has **physically-different attribute numbers from
root partition's due to dropped columns**. In this case the function creates the
`ri_PartCheckMap` map in target (child) `ResultRelInfo`, which stores correspondance
between target partition attributes and parent's. And because of that map, the
partition selection goes smoothly inside the function.

The proposed solution is to build `ri_PartCheckMap` for all the cases when the
leaf partition is modified and parent relation has dropped columns (or child
has and parent doesn't), and when the partition selection is expected. The
logic, which checks the matching of leaf's and parent's relation descriptors and
builds the `ri_PartCheckMap` is moved to a separate function
`makePartitionCheckMap`. This new function is called inside the `ExecModifyTable`
and `ExecDML` functions under different conditions.

For legacy planner this function is called for any command, except DELETE, no
matter which relation is initially specified as target (can be either root or
leaf). For example, when executing UPDATE (either split or common update) on
root relation, the planner builds the plan via `inheritance_planner`, and,
therefore, the each subplan of ModifyTable node will return the tuple slot with
attributes structure of the specific partition (some subplans will contain
dropped attributes in targetList, some won't). Here call of
`makePartitionCheckMap` is also required. For INSERT command it will be always
called as well, because `makePartitionCheckMap` validates whether oid of result
relation is equal to parent't partition oid, and this check allows us to call
the function on root relation too (if it's true, the map is not built, because
it's safe to perform operation through the root relation).

For ORCA optimizer the `makePartitionCheckMap` is called inside the `ExecDML`
function only in the case if the leaf partition is being modified. Otherwise, if
the command is called on root relation, the tuple will always have the
parent's structure, which does not break the partition selection procedure.

This patch also changes the `slot_get_partition` function, which performs the
partition selection. This function is extended with the alternative variant of
the attribute values extraction from the given slot. When the
`makePartitionCheckMap` function have created the `ri_PartCheckMap` of the child
ResultRelInfo, the attributes values array is built via that map
(inside `reconstructTupleValues`), what allows to perform the partition selection
from parent's view.

Finally, this patch additionally solves one more issue, that is not directly
related partition selection problem described above. Let's consider the
case when the parent relation has two partitions, and one of them has
incompatible attribute structure with parent's due to dropped columns. The
issue occured when the split UPDATE on parent partition was executed, and its
plan is built by legacy planner. That means that ModifyTable node contains
several sublans and several result relations inside the
`estate->es_result_relations` (inheritance planner creates them and the update is
executed for each of them). And during execution of this plan, when update is
executed on the partition, which has the same relation descriptor with parent's,
the invalid behaviour could occur when preparing the tuple for insertion (see
the last test case in tests section). In this case the partition selection at
insert stage works correct, because the partition does not differ from parent
relation. However, inside the `get_part` function, after the partition was
selected correctly from `selectPartition` function, the `targetid_get_partition`
function could wrongly form the final ResultRelInfo by creating unnecessary
`ri_partInsertMap` map, which is used in `reconstructMatchingTupleSlot` function
(inside the ExecInsert) to adjust the tuple to the selected partition descriptor
if the tuple does not fit. `targetid_get_partition` function makes the desicion on
building that map based on check, that compares two relation's descriptors:
one, which is considered to be parent's
(`parentInfo = estate->es_result_relations`), and one, which corresponds to
selected child partition (`childInfo`). But in case when UPDATE is formed by
inheritance planner, the parentInfo does not correspond the true parent
relation. It corresponds to one of other partitions. And in the case, when
selected partition is valid (it has the same relation descriptor with parent's),
and parentInfo corresponds to invalid partition (it does not match with parent
descriptor), the `ri_partInsertMap` could be built by mistake, what led to
unnecessary tuple reformatting and invalid results.

This patch solves this last issue by adding a check inside the
`targetid_get_partition` function, which ensures that the
`estate->es_result_relations` corresponds to the true root relation. That
would be true for all the cases when the DML is executed on parent relation,
except the UPDATE by legacy planner.  And if the check is passed, the
`ri_partInsertMap` is built, otherwise it's not. Moreover, during such UPDATE
there is completely no need to reconstruct the tuple, because each ModifyTable's
subplan will give already valid tuple.

bimboterminator1 committed:
Fix partition selection during DML execution when dropped columns are present

Current partitioning design faced a problem, which is connected with the
differences in physical attribute layouts of the child partition and the parent
relation in case when the parent relation has dropped columns and the newly
added child partition has not, and vice versa.

When executing modifying DML operations on a leaf partition of a partitioned
table, the executor anyway performs the partition selection in case of INSERT,
UPDATE for legacy planner and in case of INSERT, UPDATE, DELETE for ORCA
optimizer. This is done in selectPartition function, which is called from
slot_get_partition function (in case of modifying DML commands). This procedure
is done based on 1) attribute values of the tuple, which have been retrieved
from the respective subplan of ModifyTable (DML) node and whose attribute
structure corresponds the leaf partition's. 2) the root partition's attribute
numbers and partition rules, which are based on the parent (root) relation's
tuple descriptor. Thus, if the parent relation's tuple descriptor contained
dropped columns and leaf partition's did not (or visa versa, the leaf partition
had dropped columns and parent had not), the partition selection could go wrong
and lead to various execution errors.

Let's consider the following case. The partitioned table was created, and at
one moment it was decided to drop several irrelevant columns from it. After that
some new partitions were added or some were exchanged with new relations. In
this situation these newly added (exchanged) partitions don't have dropped
columns in it's structure, but the original parent relation has such dropped
attributes. Then the INSERT command into the new leaf partition is executed. In
this case for both planners the ExecInsert function is called, which then calls
slot_get_partition function in order to validate that the tuple values
correspond to specified leaf partition. And here the tuple descriptor does not
have dropped attributes, but the partition selection is performed from parent's
view, which have those attributes, and the parent's partitioning attribute
numbers are used to validate the tuple. Because of that the partition selection
procedure could either select wrong target partition (the wrong attribute was
taken from the slot values to check the partition rule) or fail to find the
partition at all, becasue none of the partition rules were satisfied. The same
issue occured when we didn't drop anything from parent relation, however we
exhanged a partition to the one with dropped columns. When inserting into
exchanged leaf partition the partition selection could also go wrong by the same
reasons (selection is performed from parent's view).

Similar issue occured for both planners in other commands, when partition
selection was performed. And in order to prevent the potentially wrong partition
selection, when modifying the leaf partition, this patch proposes the
following solution.

The legacy planner have checkPartitionUpdate function, which is called in UPDATE
cases and which checks that the tuple does not switch its partition at
update. And to our astonishment this function is able to handle the case when
the partition being modified has physically-different attribute numbers from
root partition's due to dropped columns. In this case the function creates the
ri_PartCheckMap map in target (child) ResultRelInfo, which stores correspondance
between target partition attributes and parent's. And because of that map, the
partition selection goes smoothly inside the function.

The proposed solution is to build ri_PartCheckMap for all the cases when the
leaf partition is modified and parent relation has dropped columns (or child
has and parent doesn't), and when the partition selection is expected. The
logic, which checks the matching of leaf's and parent's relation descriptors and
builds the ri_PartCheckMap is moved to a separate function
makePartitionCheckMap. This new function is called inside the ExecModifyTable
and ExecDML functions under different conditions.

For legacy planner this function is called for any command, except DELETE, no
matter which relation is initially specified as target (can be either root or
leaf). For example, when executing UPDATE (either split or common update) on
root relation, the planner builds the plan via inheritance_planner, and,
therefore, the each subplan of ModifyTable node will return the tuple slot with
attributes structure of the specific partition (some subplans will contain
dropped attributes in targetList, some won't). Here call of
makePartitionCheckMap is also required. For INSERT command it will be always
called as well, because makePartitionCheckMap validates whether oid of result
relation is equal to parent't partition oid, and this check allows us to call
the function on root relation too (if it's true, the map is not built, because
it's safe to perform operation through the root relation).

For ORCA optimizer the makePartitionCheckMap is called inside the ExecDML
function only in the case if the leaf partition is being modified. Otherwise, if
the command is called on root relation, the tuple will always have the
parent's structure, which does not break the partition selection procedure.

This patch also changes the slot_get_partition function, which performs the
partition selection. This function is extended with the alternative variant of
the attribute values extraction from the given slot. When the
makePartitionCheckMap function have created the ri_PartCheckMap of the child
ResultRelInfo, the attributes values array is built via that map
(inside reconstructTupleValues), what allows to perform the partition selection
from parent's view.

Finally, this patch additionally solves one more issue, that is not directly
related partition selection problem described above. Let's consider the same
case when the parent relation has two partitions, and one of them has
incompatible attribute structure from parent's due to dropped columns. The
issue occured when the split UPDATE on parent partition was executed. Its plan
is built by legacy planner, what means that ModifyTable node contains several
sublans and several result relations inside the estate->es_result_relations
(inheritance planner creates them and the update is executed for each of them).
And during execution of this plan, when it's turn of the "healthy" partition
to be updated (partition which has the same relation descriptor with parent's),
the invalid behaviour could occur when preparing the tuple for insertion (see
the last test in tests section). In this case the partition selection at insert
stage works correct, because the partition does not differ from parent relation.
However, inside the get_part function, after the partition was selected
correctly from selectPartition function, the targetid_get_partition function
could wrongly form the final ResultRelInfo by creating unnecessary
ri_partInsertMap map, which is used in reconstructMatchingTupleSlot function
(inside the ExecInsert) to adjust the tuple to the selected partition descriptor
if the tuple does not fit. targetid_get_partition function makes the desicion on
building that map based on check, that comapres two relation's descriptors:
one, which is considered to be parent's
(parentInfo = estate->es_result_relations), and one, which corresponds to
selected child partition (childInfo). But in case then UPDATE is formed by
inheritance planner, the parentInfo does not correspond the true parent relation
It corresponds to one of other partitions. And in the case, when selected
partition is "healthy" (which has the same relation descriptor with parent's),
and parentInfo corresponds to "unhealthy" partition, the ri_partInsertMap could
be built by mistake, what led to unnecessary tuple reformatting and invalid
results.

This patch solves this last issue by adding a check inside the
targetid_get_partition function, which ensures that the selected partition
is not the same as in estate->es_result_relation_info, which will be the same
in case of split UPDATE by legacy planner. And if the check is passed, the
ri_partInsertMap is built, otherwise it's not. Moreover, during such UPDATE
there is completely no need to reconstruct the tuple, because each ModifyTable's
subplan will give already valid tuple.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/56288

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/693944

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/56748

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/56749

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/711280

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57096

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/723945

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/723946

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/723944

KnightMurloc commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2448,79 @@ ExecSquelchModifyTable(ModifyTableState *node)
+			/* And save it for later use. */
+			resultRelInfo->ri_PartCheckMap = map;
+
+			resultRelInfo->ri_PartCheckTupDescMatch = -1;
+		}
+		else
+			resultRelInfo->ri_PartCheckTupDescMatch = 1;
```
why `ri_PartCheckTupDescMatch` can contain three values -1, 0 and 1 when we check it only for 0? should it be a bool?

bimboterminator1 replied:
I 've just moved out the map creation logic from the `checkPartitionUpdate` [function](https://github.com/arenadata/gpdb/blob/c1645466534bedbc12ac52bca133f0ff73576636/src/backend/executor/nodeModifyTable.c#L1127). I didn't want to refactor or simplify the map-related infrastructure.
`0` indicates that for the given ResultRelation the descriptor matching check haven't been perfomed.
`+1`, `-1` could be useful to indicate whether ResultRelation's descriptor matches with the parent partition's or not. I agree, currently the `0` check is used, but someone may need that in the future, and it would be more convinient to monitor the matching through 3 states.

KnightMurloc commented:
why should we call `makePartitionCheckMap` for delete in the case of ORCA and may not do it in the case of postgres?

bimboterminator1 commented:
> why should we call `makePartitionCheckMap` for delete in the case of ORCA and may not do it in the case of postgres?

Planner's DELETE does not select patitions.

https://github.com/arenadata/gpdb/blob/c1645466534bedbc12ac52bca133f0ff73576636/src/backend/executor/nodeModifyTable.c#L677

KnightMurloc commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2448,79 @@ ExecSquelchModifyTable(ModifyTableState *node)
+	 */
+	parentRelid = estate->es_result_partitions->part->parrelid;
+
+	if (resultRelInfo->ri_PartitionParent &&
+		parentRelid != RelationGetRelid(resultRelInfo->ri_PartitionParent))
+	{
+		resultRelInfo->ri_PartCheckTupDescMatch = 0;
+		if (resultRelInfo->ri_PartCheckMap != NULL)
+			pfree(resultRelInfo->ri_PartCheckMap);
+		if (resultRelInfo->ri_PartitionParent)
+			relation_close(resultRelInfo->ri_PartitionParent, AccessShareLock);
+	}
```
in which case is this logic executed? even when working with a subpartition, parentRelid will refer to the root table.

bimboterminator1 replied:
Probably it will never be executed. (There is no case, cause the `estate->es_result_partitions->part->parrelid` can't be changed during execution of ModifyTable. And the `ri_PartitionParent` is created only in that function and it can't be rewritten later) I'll remove that overhead code.

KnightMurloc commented on src/backend/executor/execMain.c:
```diff
@@ -4603,10 +4604,27 @@ targetid_get_partition(Oid targetid, EState *estate, bool openIndices)
+		 * parentInfo represents another partition, which is not the true
+		 * parent).
+		 */
+		if (!currRelInfo || RelationGetRelid(currRelInfo->ri_RelationDesc) != targetid)
```
in which case can `currRelInfo` be null and why should we do `map_part_attrs` in this case?

bimboterminator1 replied:
Yep, there is no case for it to be NULL. However, after short exploration the usages of `targetid_get_partition` function (it is also used in case of COPY cmd), I can conclude that the issue related to these changes is only for UPDATE command for planner case (for inheritance case). Therefore, I'll reconsider that part of code and probably move it to another place.

bimboterminator1 replied:
If we want to prevent the `ri_partInsertMap` construction, we can only make changes inside the `targetid_get_partition` function. Otherwise, for example, if we make such validation after the partition was selected (at execInsert), there would be no option except the `pfree(resultRelInfo->ri_partInsertMap), resultRelInfo->ri_partInsertMap = NULL ` call.

Therefore, I've just changed the map building condition, which allows to create the map only when the DML is executed on root partition.

bimboterminator1 committed:
Remove validation overhead

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57296

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/732029

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/732030

bimboterminator1 committed:
Change condition for attribute mapping

bimboterminator1 committed:
Fix partition selection during DML execution when dropped columns are present

Current partitioning design faced a problem, which is connected with the
differences in physical attribute layouts of the child partition and the parent
relation in case when the parent relation has dropped columns and the newly
added child partition has not, and vice versa.

When executing modifying DML operations on a leaf partition of a partitioned
table, the executor anyway performs the partition selection in case of INSERT,
UPDATE for legacy planner and in case of INSERT, UPDATE, DELETE for ORCA
optimizer. This is done in selectPartition function, which is called from
slot_get_partition function (in case of modifying DML commands). This procedure
is done based on 1) attribute values of the tuple, which have been retrieved
from the respective subplan of ModifyTable (DML) node and whose attribute
structure corresponds the leaf partition's. 2) the root partition's attribute
numbers and partition rules, which are based on the parent (root) relation's
tuple descriptor. Thus, if the parent relation's tuple descriptor contained
dropped columns and leaf partition's did not (or visa versa, the leaf partition
had dropped columns and parent had not), the partition selection could go wrong
and lead to various execution errors.

Let's consider the following case. The partitioned table was created, and at
one moment it was decided to drop several irrelevant columns from it. After that
some new partitions were added or some were exchanged with new relations. In
this situation these newly added (exchanged) partitions don't have dropped
columns in it's structure, but the original parent relation has such dropped
attributes. Then the INSERT command into the new leaf partition is executed. In
this case for both planners the ExecInsert function is called, which then calls
slot_get_partition function in order to validate that the tuple values
correspond to specified leaf partition. And here the tuple descriptor does not
have dropped attributes, but the partition selection is performed from parent's
view, which have those attributes, and the parent's partitioning attribute
numbers are used to validate the tuple. Because of that the partition selection
procedure could either select wrong target partition (the wrong attribute was
taken from the slot values to check the partition rule) or fail to find the
partition at all, becasue none of the partition rules were satisfied. The same
issue occured when we didn't drop anything from parent relation, however we
exhanged a partition to the one with dropped columns. When inserting into
exchanged leaf partition the partition selection could also go wrong by the same
reasons (selection is performed from parent's view).

Similar issue occured for both planners in other commands, when partition
selection was performed. And in order to prevent the potentially wrong partition
selection, when modifying the leaf partition, this patch proposes the
following solution.

The legacy planner have checkPartitionUpdate function, which is called in UPDATE
cases and which checks that the tuple does not switch its partition at
update. And to our astonishment this function is able to handle the case when
the partition being modified has physically-different attribute numbers from
root partition's due to dropped columns. In this case the function creates the
ri_PartCheckMap map in target (child) ResultRelInfo, which stores correspondance
between target partition attributes and parent's. And because of that map, the
partition selection goes smoothly inside the function.

The proposed solution is to build ri_PartCheckMap for all the cases when the
leaf partition is modified and parent relation has dropped columns (or child
has and parent doesn't), and when the partition selection is expected. The
logic, which checks the matching of leaf's and parent's relation descriptors and
builds the ri_PartCheckMap is moved to a separate function
makePartitionCheckMap. This new function is called inside the ExecModifyTable
and ExecDML functions under different conditions.

For legacy planner this function is called for any command, except DELETE, no
matter which relation is initially specified as target (can be either root or
leaf). For example, when executing UPDATE (either split or common update) on
root relation, the planner builds the plan via inheritance_planner, and,
therefore, the each subplan of ModifyTable node will return the tuple slot with
attributes structure of the specific partition (some subplans will contain
dropped attributes in targetList, some won't). Here call of
makePartitionCheckMap is also required. For INSERT command it will be always
called as well, because makePartitionCheckMap validates whether oid of result
relation is equal to parent't partition oid, and this check allows us to call
the function on root relation too (if it's true, the map is not built, because
it's safe to perform operation through the root relation).

For ORCA optimizer the makePartitionCheckMap is called inside the ExecDML
function only in the case if the leaf partition is being modified. Otherwise, if
the command is called on root relation, the tuple will always have the
parent's structure, which does not break the partition selection procedure.

This patch also changes the slot_get_partition function, which performs the
partition selection. This function is extended with the alternative variant of
the attribute values extraction from the given slot. When the
makePartitionCheckMap function have created the ri_PartCheckMap of the child
ResultRelInfo, the attributes values array is built via that map
(inside reconstructTupleValues), what allows to perform the partition selection
from parent's view.

Finally, this patch additionally solves one more issue, that is not directly
related partition selection problem described above. Let's consider the
case when the parent relation has two partitions, and one of them has
incompatible attribute structure with parent's due to dropped columns. The
issue occured when the split UPDATE on parent partition was executed, and its
plan is built by legacy planner. That means that ModifyTable node contains
several sublans and several result relations inside the
estate->es_result_relations (inheritance planner creates them and the update is
executed for each of them). And during execution of this plan, when update is
executed on the partition, which has the same relation descriptor with parent's,
the invalid behaviour could occur when preparing the tuple for insertion (see
the last test case in tests section). In this case the partition selection at
insert stage works correct, because the partition does not differ from parent
relation. However, inside the get_part function, after the partition was
selected correctly from selectPartition function, the targetid_get_partition
function could wrongly form the final ResultRelInfo by creating unnecessary
ri_partInsertMap map, which is used in reconstructMatchingTupleSlot function
(inside the ExecInsert) to adjust the tuple to the selected partition descriptor
if the tuple does not fit. targetid_get_partition function makes the desicion on
building that map based on check, that compares two relation's descriptors:
one, which is considered to be parent's
(parentInfo = estate->es_result_relations), and one, which corresponds to
selected child partition (childInfo). But in case when UPDATE is formed by
inheritance planner, the parentInfo does not correspond the true parent
relation. It corresponds to one of other partitions. And in the case, when
selected partition is valid (it has the same relation descriptor with parent's),
and parentInfo corresponds to invalid partition (it does not match with parent
descriptor), the ri_partInsertMap could be built by mistake, what led to
unnecessary tuple reformatting and invalid results.

This patch solves this last issue by adding a check inside the
targetid_get_partition function, which ensures that the
estate->es_result_relations corresponds to the true root relation. That
would be true for all the cases when the DML is executed on parent relation,
except the UPDATE by legacy planner.  And if the check is passed, the
ri_partInsertMap is built, otherwise it's not. Moreover, during such UPDATE
there is completely no need to reconstruct the tuple, because each ModifyTable's
subplan will give already valid tuple.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57473

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/740256

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/740257

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/740254

KnightMurloc commented on src/backend/executor/execMain.c:
```diff
@@ -4603,10 +4604,27 @@ targetid_get_partition(Oid targetid, EState *estate, bool openIndices)
+		 * inheritance plan execution, the tuple descriptor already matches
+		 * the partition's, and the extra mapping is unnecessary.
+		 */
+		if (RelationGetRelid(parentInfo->ri_RelationDesc) == parentRelid)
+			map_part_attrs(parentInfo->ri_RelationDesc,
+						   childInfo->ri_RelationDesc,
+						   &(childInfo->ri_partInsertMap),
+						   TRUE); /* throw on error, so result not needed */
```
this logic handles the insert into root partition. right? why don't we check if we need ri_partInsertMap or not? after all, if there are no deleted columns, then the map is unnecessary (right?). this check is done in makePartitionCheckMap.
https://github.com/arenadata/gpdb/blob/f6aa64215a15482e98fbd19b0dc73e79a78604af/src/backend/executor/nodeModifyTable.c#L2495

bimboterminator1 replied:
Well, if there are no dropped columns, the `map_part_attrs` simply won't build the ` ri_partInsertMap`, because the `parentInfo->ri_RelationDesc` and  `childInfo->ri_RelationDesc` are equal.  Again, when we are working through root partition, we definetely can be sure that the tuple corresponds to parent structure (except the UPDATE from legacy planner). That means that we will need the ` ri_partInsertMap` if after partition selection the child partition has different descriptor from parent's, and we need to map parent attributes to child's.

KnightMurloc commented on src/test/regress/sql/qp_dropped_cols.sql:
```diff
@@ -8687,3 +8687,114 @@ DELETE FROM dist_key_dropped_pt WHERE b=6;
+UPDATE t_part_dropped_1_prt_p2 SET c1 = 2;
+UPDATE t_part_dropped SET c1 = 3;
+
+INSERT INTO t_part_dropped VALUES (1, 2, 0);
+
+-- Ensure that split update on leaf partition does not throw constraint error
+-- (legacy planner does not choose the wrong partition at insert) and does not
+-- throw partition selection error (ORCA).
+UPDATE t_part_dropped_1_prt_p2 SET c1 = 2 WHERE c4 = 0;
```
Is this insert is a preparation for the below test? Maybe then move it under the comment so that it is clear what it refers to? same on the line 8757

bimboterminator1 replied:
Done

KnightMurloc commented on src/test/regress/sql/qp_dropped_cols.sql:
```diff
@@ -8687,3 +8687,114 @@ DELETE FROM dist_key_dropped_pt WHERE b=6;
+-- plan (legacy planner). Ensure that split update does not reconstruct the
+-- tuple at insert.
+INSERT INTO t_part VALUES (1, 4, 2, 2);
+UPDATE t_part SET c1=3;
```
add spaces between = so that the style matches the rest of the updates in your tests.

bimboterminator1 replied:
Done

bimboterminator1 committed:
change variable assignent

bimboterminator1 committed:
Change test formatting and comments

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57643

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/748758

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/748759

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/748752

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2448,69 @@ ExecSquelchModifyTable(ModifyTableState *node)
+			resultRelInfo->ri_PartCheckTupDescMatch = -1;
+		}
+		else
+			resultRelInfo->ri_PartCheckTupDescMatch = 1;
```
Why not all comments and code were moved here from checkPartitionUpdate? For example
```c
resultRelInfo->ri_PartCheckTupDescMatch = 0;
```
and/or
```c
	/*
	 * Check this at the first pass only to avoid repeated catalog access.
	 */
```
and so on

bimboterminator1 replied:
Have resotred omitted code and comments

RekGRpth commented on src/test/regress/sql/qp_dropped_cols.sql:
```diff
@@ -8687,3 +8687,112 @@ DELETE FROM dist_key_dropped_pt WHERE b=6;
+SELECT count(*) FROM t_part_dropped_1_prt_p2;
+SELECT * FROM t_part_dropped_1_prt_p0;
+
+-- For ORCA the partition selection error should not occur.
```
How can you be sure that the plan was generated by ORCA?

bimboterminator1 replied:
Have added explain output for several commands.

InnerLife0 commented:
Generally looks OK, but I want to talk about test coverage after your vacation.

bimboterminator1 committed:
Restore omitted code and comments

bimboterminator1 committed:
Extented tests with explain parts

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/58776

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/803359

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/803358

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/803360

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/803354

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4560,6 +4560,7 @@ targetid_get_partition(Oid targetid, EState *estate, bool openIndices)
 	ResultRelInfo *childInfo = estate->es_result_relations;
 	ResultPartHashEntry *entry;
 	bool		found;
+	Oid			parentRelid = estate->es_result_partitions->part->parrelid;
```
Could you reduce the scope of this variable by moving it to the block where it is used?

bimboterminator1 replied:
Done

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4631,22 +4649,66 @@ values_get_partition(Datum *values, bool *nulls, TupleDesc tupdesc,
+	}
+	else
+	{
+		max_attr = estate->es_partition_state->max_partition_attr;
```
```suggestion
		AttrNumber max_attr = estate->es_partition_state->max_partition_attr;
```

bimboterminator1 replied:
Done

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4631,22 +4649,66 @@ values_get_partition(Datum *values, bool *nulls, TupleDesc tupdesc,
 {
-	ResultRelInfo *resultRelInfo;
+	ResultRelInfo *resultRelInfo = estate->es_result_relation_info;
+	TupleDesc	tupdesc = NULL;
```
nullify is redundant

bimboterminator1 replied:
Done

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4631,22 +4649,66 @@ values_get_partition(Datum *values, bool *nulls, TupleDesc tupdesc,
+		Datum	   *parent_values;
+		bool	   *parent_nulls;
+		Relation	parentRel = resultRelInfo->ri_PartitionParent;
+		TupleDesc	parentTupdesc;
```
this variable is redundant

bimboterminator1 replied:
Done

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4631,22 +4649,66 @@ values_get_partition(Datum *values, bool *nulls, TupleDesc tupdesc,
+							   parent_values, parent_nulls, parentTupdesc->natts);
+
+		/* Now we have values/nulls in parent's view. */
+		values = parent_values;
```
this line are redundant if you swap parent_values and values before

bimboterminator1 replied:
"Before" means where?

RekGRpth replied:
before current line

bimboterminator1 replied:
I don't get it, what does "swap" means? totally remove parent_values and parent_nulls?
```		
values = palloc(tupdesc->natts * sizeof(Datum));
nulls = palloc0(tupdesc->natts * sizeof(bool));

map = resultRelInfo->ri_PartCheckMap;
reconstructTupleValues(map, slot_get_values(slot), slot_get_isnull(slot), slot->tts_tupleDescriptor->natts,
					values, nulls, tupdesc->natts);
```

RekGRpth replied:
> values = palloc(tupdesc->natts * sizeof(Datum));
> nulls = palloc0(tupdesc->natts * sizeof(bool));

yes



> reconstructTupleValues(map, slot_get_values(slot), slot_get_isnull(slot), slot->tts_tupleDescriptor->natts,
> 					values, nulls, tupdesc->natts);

```c

		reconstructTupleValues(map, some_values, some_nulls, slot->tts_tupleDescriptor->natts,
							   values, nulls, parentTupdesc->natts);
```

bimboterminator1 replied:
Done

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4631,22 +4649,66 @@ values_get_partition(Datum *values, bool *nulls, TupleDesc tupdesc,
+
+		/* Now we have values/nulls in parent's view. */
+		values = parent_values;
+		nulls = parent_nulls;
```
this line are redundant if you swap parent_nulls and nulls before

bimboterminator1 replied:
Done

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -85,6 +85,39 @@ ExecDML(DMLState *node)
+		 */
+		if (RelationGetRelid(relInfo->ri_RelationDesc) !=
+			node->ps.state->es_result_partitions->part->parrelid)
+			makePartitionCheckMap(node->ps.state,
```
Why you split this line?

bimboterminator1 replied:
Fixed

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2449,86 @@ ExecSquelchModifyTable(ModifyTableState *node)
+		parentRelid != RelationGetRelid(resultRelInfo->ri_RelationDesc))
+	{
+		Relation	parentRel;
+		TupleDesc	resultTupdesc,
```
```suggestion
		TupleDesc	resultTupdesc, parentTupdesc;
```

bimboterminator1 replied:
pgindent shows different formating (as it is presented in current variant).

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2449,86 @@ ExecSquelchModifyTable(ModifyTableState *node)
+		parentTupdesc = RelationGetDescr(parentRel);
+		if (!equalTupleDescs(resultTupdesc, parentTupdesc, false))
+		{
+			AttrMap    *map;
```
```suggestion
			AttrMap		   *map;
```

bimboterminator1 replied:
pgindent shows different formating (as it is presented in current variant).

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2449,86 @@ ExecSquelchModifyTable(ModifyTableState *node)
+		if (!equalTupleDescs(resultTupdesc, parentTupdesc, false))
+		{
+			AttrMap    *map;
+			MemoryContext oldcontext;
```
```suggestion
			MemoryContext	oldcontext;
```

bimboterminator1 replied:
pgindent shows different formating (as it is presented in current variant).

bimboterminator1 committed:
Refactoring

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -4631,22 +4649,61 @@ values_get_partition(Datum *values, bool *nulls, TupleDesc tupdesc,
 							 true);
 
+	/* Free up if we allocated mapped attributes. */
+	if (values != slot_get_values(slot) &&
```
How about
```c
if (values != slot_get_values(slot))
    pfree(values);

if (nulls != slot_get_isnull(slot))
    pfree(nulls);
```

bimboterminator1 replied:
It's more beautiful, ok

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/828762

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/828761

bimboterminator1 committed:
extra refactoring

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/829093

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/829094

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/829095

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/829088

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2449,86 @@ ExecSquelchModifyTable(ModifyTableState *node)
+	parentRelid = estate->es_result_partitions->part->parrelid;
+
+	/*
+	 * I don't believe this is the case currently, but we check the parent relid
```
line overflows 80

bimboterminator1 replied:
fixed

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2506,3 +2449,86 @@ ExecSquelchModifyTable(ModifyTableState *node)
+
+	/*
+	 * I don't believe this is the case currently, but we check the parent relid
+	 * in case the updating partition has changed since the last time we opened it.
```
line overflows 80

bimboterminator1 replied:
fixed

bimboterminator1 committed:
comment 80 alignment

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/59339

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/830048

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/830049

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/830051

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4207

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/59739

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/850319

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/850320
